### DISCUSSION TITLE
This can/should co-exist with alias-tips

### DISCUSSION THREAD

**Discussion by sagarkar10** [olets/zsh-abbr#51] [Ideas] [answered]
That would be awesome if we can skip `alias-tips` [repo](https://github.com/djui/alias-tips) suggestions if the command is coming from `zsh-abbr`
Does it make sense to use both? Thoughts?

&nbsp;

_edit from olets:_

> [!NOTE]
> Answered at https://github.com/olets/zsh-abbr/discussions/51#discussioncomment-1855440

**Comment by olets** [comment, marked as answer]
Thanks for the idea!

alias-tips has an excludes array (https://github.com/djui/alias-tips#exclude-some-aliases). If for every abbreviation you have an alias with the same short form (for example, `alias x="my command"` and `abbr x="my command"`) then adding this snippet to your `.zshrc` should automatically exclude all abbreviations when the shell initializes:

```shell
export ZSH_PLUGINS_ALIAS_TIPS_EXCLUDES="${(f)$(abbr l)}"
```

(that is, a single string from `abbr l`'s multi-line output). If you're already excluding things,

```shell
export ZSH_PLUGINS_ALIAS_TIPS_EXCLUDES="myalias ${(f)$(abbr l)}"
```

After adding a new abbreviation, open a new terminal and/or refresh open terminals (`source ~/.zshrc` or, preferably, `exec zsh`) to exclude the newly added abbreviation.

Does that meet your need?

**Comment by sagarkar10** [comment]
Yes, that's exactly what I needed! Feeling stupid, didn't figure out myself!
Thanks a lot!

**Comment by sagarkar10** [comment]
Also, a 2nd thought, we should auto sync `alias` and `abbr` in-case people are using both and transitioning!

**Comment by olets** [comment]
I've chosen to not support _auto_ syncing because it would require letting abbr write and possibly first parse `.zshrc`. Have you tried [`abbr export-aliases`](https://zsh-abbr.olets.dev/commands.html#export-aliases) and [`abbr import-aliases`](https://zsh-abbr.olets.dev/commands.html#import-aliases)?

You could do something like

1. run `abbr import-aliases` once
2. add to the end of your `.zshrc`

    ```shell
    abbr export-aliases > $HOME/aliases-from-abbr
    source $HOME/aliases-from-abbr
    ```

Putting it at the very end of your `.zshrc` will keep the aliases from breaking other parts of your shell initialization.

//

Another approach: run

```shell
abbr import-aliases
```

once and then delete as many aliases as possible. That's my personal recommendation. It would probably even obviate your alias-tips problem. If you decide to switch back to aliases, run `abbr export-aliases` once, and copy and paste that into your shell config.

**Comment by sagarkar10** [comment]
I understand both the workarounds you proposed!
But I wanted to suggest something different
Assuming all the aliases are present in `.zsh-aliases` which is sourced in `.zshrc`

Someone might want to use both `abbr` and `alias` interchangeably.

What I suggest is 
- we create a hook that runs `abbr export-aliases` to `.zsh-aliases` file at session end (or on-demand)
- Another hook that clears (`abbr clear-all`) `abbr`  and then add `abbr import-aliases`

Let me know if that makes sense, I can add a PR, or we can discuss in more details about the implementation!

**Comment by olets** [comment]
Interesting idea! For now, it doesn't feel to me like syncing abbreviations and aliases is something zsh-abbr should take on — it's sort of a meta responsibility. But let's keep thinking it through.

//

> `abbr clear-all`

Now and then I consider making a "delete all abbreviations" function… But I'm afraid if it existed I'd run it by mistake! The easiest way to roll your own will be

```shell
echo > $ABBR_USER_ABBREVIATIONS_FILE
abbr load
```

>  at session end

I'm not aware of a way to do this, but sounds cool. Are you?

//

What if you invert my previous idea, letting aliases be the source of truth

```shell
# .zshrc

source .zsh-aliases
# load zsh-abbr and then
abbr import-aliases --quiet
```

and to add an alias and an abbreviation use

```shell
% echo alias x=y > .zsh-aliases && exec zsh
```

(Could make a function for that, and `abbralias x=y`.) Or am I still not seeing your use case?

**Reply by olets** [reply]
Huh. That's my first time converting an issue to a discussion. Expected it to migrate as a single thread, and wasn't thinking about how discussions don't show up in the Issues tab search results.

**Reply by sagarkar10** [reply]
About the end session thing trap can work I believe. [Here](https://stackoverflow.com/a/18222158)

**Reply by olets** [reply]
Let me know if you get it working!

**Comment by mainrs** [comment]
I stumbled upon this on my search to see if this plugin supports abbreviation hints like alias-tips does. I.e., I write the full command but there already exists an abbreviation for it. So the plugin shows me at the end of the output the abbreviation I could have used.

**Reply by olets** [reply]
It doesn't. After a quick read through alias-tips and zsh-you-should-use, it looks to me like it'd be easier to add abbreviation support to zsh-you-should-use than to alias-tips.

(Not a solution exactly, but for a quick reminder of all your abbreviations run `abbr`.)

**Reply by p1r473** [reply]
I'm working on it as we speak. Need a few more days and it's done.
https://github.com/MichaelAquilina/zsh-you-should-use/pull/131

**Reply by olets** [reply]
This is now available in v6 https://zsh-abbr.olets.dev/reminders.html
